sklearn/metrics/scorer.py

```diff
@@ -272,14 +274,122 @@ def check_scoring(estimator, scoring=None, allow_none=False):
+
+
+def _check_multimetric_scoring(estimator, scoring=None, allow_none=False):
+    """Check the scoring parameter in cases when multiple metrics are allowed
+
+    Parameters
+    ----------
+    estimator : sklearn estimator instance
+        The estimator for which the scoring will be applied.
+
+    scoring : string, callable or None, optional, default: None
+        A single string (see :ref:`_scoring_parameter`) or a callable
+        (see :ref:`_scoring`) to evaluate the predictions on the test set.
+
+        For evaluating multiple metrics either give a list of unique strings
+        or a dict with names as keys and callables as values.
+
+        NOTE that when using custom scorers, each scorer should return a single
+        value. Single scorers returning a list/array of values may be wrapped
+        into multiple scorers that return one value each.
+
+        If ``None``, the estimator's default scorer, is used.
+        The return value in that case will be ``{'score': <default_scorer>}``.
+        If the estimator's default scorer is not available, a ``TypeError``
+        is raised.
+
+    allow_none : boolean, default: False
+        Whether to allow None even if estimator does not have a score method.
+
+    Returns
+    -------
+    scorers_dict : dict
+        A dict mapping each scorer name to its validated scorer.
+
+        The validated scorer could be None if ``allow_none`` is ``True``
+        (``{'score': None}``).
+
+        The scorer_name is set to ``'score'`` if the default scorer is
+        used.
+
+    is_multimetric : bool
+        True if scorer is a list/tuple or dict of callables
+        False if scorer is None/str/callable
+    """
+    if callable(scoring) or scoring is None or isinstance(scoring,
+                                                          six.string_types):
+        scorers = {"score": check_scoring(estimator, scoring=scoring,
+                                          allow_none=allow_none)}
+        return scorers, False
     else:
-        raise TypeError(
-            "If no scoring is specified, the estimator passed should "
-            "have a 'score' method. The estimator %r does not." % estimator)
+        err_msg_generic = ("scoring should either be a single string or "
+                           "callable for single metric evaluation or a "
+                           "list/tuple of strings or a dict of scorer name "
+                           "mapped to the callable for multiple metric "
+                           "evaluation. Got %s of type %s"
+                           % (repr(scoring), type(scoring)))
+
+        if isinstance(scoring, (list, tuple, set)):
+            err_msg = ("The list/tuple elements must be unique "
+                       "strings of predefined scorers. ")
+            invalid = False
+            try:
+                keys = set(scoring)
+            except TypeError:  # For list of lists
```

Comment:
Oh yes that too. It will catch all which cannot be made as a set... I should remove that comment I guess. Thx